### DISCUSSION TITLE
Getting response headers asynchronously by chunk

### DISCUSSION THREAD

**Discussion by frextrite** [hyperium/hyper#3758] [Q&A] [answered]
Is there a way to get response headers asynchronously byte by byte (chunk by chunk), the same way we can get response body [1], instead of making a blocking synchronous call [2]? Basically, I'm trying to create a timing utility that returns TTFB (time taken to receive the first byte of response) information for an HTTP request and for that I'd like to get a hook into when the first chunk of response headers is received instead of waiting until all the headers are received which could potentially take multiple reads.

[1]
For example, the following [piece of code](https://hyper.rs/guides/1/client/basic/) can be used to get response body
```rust
// Stream the body, writing each frame to stdout as it arrives
while let Some(next) = res.frame().await {
    let frame = next?;
    if let Some(chunk) = frame.data_ref() {
        io::stdout().write_all(chunk).await?;
    }
}
```

[2]
This gives us the headers in one-shot, but the `send_request(req).await` returns when all the headers have been read and parsed which could take multiple socket reads and the time elapsed could be more than the actual time to first byte.
```rust
// Await the response...
let begin = Instant::now();
let mut res = sender.send_request(req).await?;
let elapsed = begin.elapsed(); // could be more than the actual time to first byte

println!("Response status: {}", res.status());
println!("Response headers: {}", res.headers());
```

**Comment by seanmonstar** [comment, marked as answer]
There isn't a way to get the headers "by chunks", and I don't expect there will be. However, getting TTFB is a pretty common ask, and is something that we would like to expose in some sort of metrics API. We just don't have one yet, nor even a proposal. If you're interested in working on one, let me know. It will be part of the next roadmap update (though will require a champion).

**Reply by frextrite** [reply]
> something that we would like to expose in some sort of metrics API

Just wanted to clarify if hyper already has a metrics API which would be extended to add support for TTFB or is this metrics API something entirely new?

**Reply by seanmonstar** [reply]
It would be new (see https://github.com/hyperium/hyper/pull/3763).
